src/core/queryClient.ts

```diff
@@ -129,14 +130,28 @@ export class QueryClient {
 
   setQueryData<TData>(
     queryKey: QueryKey,
-    updater: Updater<TData | undefined, TData>,
+    updater: Updater<TData | undefined, TData> | undefined,
     options?: SetDataOptions
-  ): TData {
-    const parsedOptions = parseQueryArgs(queryKey)
-    const defaultedOptions = this.defaultQueryOptions(parsedOptions)
-    return this.queryCache
-      .build(this, defaultedOptions)
-      .setData(updater, options)
+  ): TData | undefined {
+    if (typeof updater === 'undefined') {
+      return undefined
+    }
+
+    let query = this.queryCache.find<TData>(queryKey)
+    const prevData = query?.state.data
+    const data = functionalUpdate(updater, prevData)
+
+    if (typeof data === 'undefined') {
+      return
+    }
+
+    if (!query) {
+      const parsedOptions = parseQueryArgs(queryKey)
+      const defaultedOptions = this.defaultQueryOptions(parsedOptions)
+      query = this.queryCache.build(this, defaultedOptions)
+    }
+
+    return query.setData(data, options)
```

Comment:
looking at this once more, I don't think we need the `if` guard:

```suggestion
const parsedOptions = parseQueryArgs(queryKey)
const defaultedOptions = this.defaultQueryOptions(parsedOptions)
return this.queryCache
  .build(this, defaultedOptions)
  .setData(data, options)
```

Reply:
Right, since we did not have the guard before.